runtime/lua/vim/cache.lua

```diff
@@ -0,0 +1,481 @@
+  if M.enabled then
+    return
+  end
+  M.enabled = true
+  vim.fn.mkdir(vim.fn.fnamemodify(M.path, ':p'), 'p')
+  -- selene: allow(global_usage)
+  _G.loadfile = Cache.loadfile
+  -- add lua loader
+  table.insert(package.loaders, 2, Cache.loader)
+  -- add libs loader
+  table.insert(package.loaders, 3, Cache.loader_lib)
+  -- remove Neovim loader
+  for l, loader in ipairs(package.loaders) do
+    if loader == vim._load_package then
+      table.remove(package.loaders, l)
+      break
+    end
+  end
+  -- TODO: add an autocmd on BufWritePost that checks if its in a /lua folder
+  -- if thats the case, then reset the plugin path.
+  -- This will make sure we can properly load new top-level lua modules
+end
+
+--- Disables the cache:
+--- * removes the cache loaders
+--- * adds the Neovim loader
+function M.disable()
+  if not M.enabled then
+    return
+  end
+  M.enabled = false
+  -- selene: allow(global_usage)
+  _G.loadfile = Cache._loadfile
+  ---@diagnostic disable-next-line: no-unknown
+  for l, loader in ipairs(package.loaders) do
+    if loader == Cache.loader or loader == Cache.loader_lib then
+      table.remove(package.loaders, l)
+    end
+  end
+  table.insert(package.loaders, 2, vim._load_package)
+end
+
+--- Return the top-level `/lua/*` modules for this path
+---@param path string path to check for top-level lua modules
+---@return string[]
+function M.lsmod(path)
+  if not Cache._indexed[path] then
+    local start = uv.hrtime()
+    Cache._indexed[path] = {}
+    local handle = vim.loop.fs_scandir(path .. '/lua')
+    while handle do
+      local name, t = vim.loop.fs_scandir_next(handle)
+      if not name then
+        break
+      end
+      -- HACK: type is not always returned due to a bug in luv
+      t = t or uv.fs_stat(path .. '/lua/' .. name).type
+      ---@type string
+      local topname
+      local ext = name:sub(-4)
+      if ext == '.lua' or ext == '.dll' then
+        topname = name:sub(1, -5)
+      elseif name:sub(-3) == '.so' then
+        topname = name:sub(1, -4)
+      elseif t == 'link' or t == 'directory' then
+        topname = name
+      end
+      if topname then
+        Cache._indexed[path][topname] = true
+        Cache._topmods[topname] = Cache._topmods[topname] or {}
+        if not vim.tbl_contains(Cache._topmods[topname], path) then
+          table.insert(Cache._topmods[topname], path)
+        end
+      end
+    end
+    M.track('lsmod', start)
+  end
+  return Cache._indexed[path]
+end
+
+--- Debug function that wrapps all loaders and tracks stats
+function M.profile_loaders()
+  for l, loader in pairs(package.loaders) do
+    local loc = debug.getinfo(loader, 'Sn').source:sub(2)
+    package.loaders[l] = function(modname)
+      local start = vim.loop.hrtime()
+      local ret = loader(modname)
+      M.track('loader ' .. l .. ': ' .. loc, start)
+      M.track('loader_all', start)
+      return ret
+    end
+  end
+end
+
+--- Prints all cache stats
+function M.inspect()
```

Comment:
And this. This should really be just for developer use.

Reply:
we will want to have visibility  `:checkhealth` , somehow

Reply:
Should probably be something separate, though; this is more like low-level debug information reporting the timings of various cache loader steps. `:checkhealth` should be more high-level ("cache is enabled and has size x Kb").

Reply:
I'd rather keep this, since it's similar to other inspect methods we have?

Reply:
I was referring to `:checkhealth` specifically; it should not use this `inspect` but something else better suited to that context.

No opinion on `inspect()` itself.

Reply:
> I'd rather keep this, since it's similar to other inspect methods we have?

+1 but mark it as private and/or "experimental" so that we don't have to think much about future-proofing it yet? I assume that was @lewis6991 's main concern (which I agree with).